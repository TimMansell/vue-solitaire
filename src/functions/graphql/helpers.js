import { gql } from 'apollo-server-lambda';

// eslint-disable-next-line import/prefer-default-export
export const grabQuery = ({ body }) => {
  const { query } = JSON.parse(body);

  const gqlQuery = gql`
    ${query}
  `;

  return gqlQuery;
};

export const grabVariables = ({ body }) => {
  const { variables } = JSON.parse(body);

  return variables;
};
