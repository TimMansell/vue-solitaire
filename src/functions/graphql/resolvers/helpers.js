import { gql } from 'apollo-boost';

export const formatVariables = (args, params) => {
  const { id, data } = args;

  const variables = {
    id,
    data: {
      ...data,
      ...params,
    },
  };

  return variables;
};

export const updateGameMutation = async (client, variables) => {
  const mutation = gql`
    mutation UpdateAGame($id: ID!, $data: GameInput!) {
      updateGame(id: $id, data: $data) {
        _id
        gameNumber
      }
    }
  `;

  const body = await client.mutate({
    mutation,
    variables,
  });

  return body.data.updateGame;
};

export const newGameMutation = async (client, variables) => {
  const mutation = gql`
    mutation NewGame($date: String!) {
      newGame(date: $date) {
        _id
        gameNumber
      }
    }
  `;

  const body = await client.mutate({
    mutation,
    variables,
  });

  return body.data.newGame;
};
