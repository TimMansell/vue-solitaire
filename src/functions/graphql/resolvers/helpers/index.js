export {
  insertIntoDb,
  countItemsInDb,
  findAllItems,
  findItemsInDb,
  findItemInDb,
} from './db';
export { findLeaderboardItems } from './find';
export { createPlayerName } from './user';
