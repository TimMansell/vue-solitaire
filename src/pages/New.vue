<template>
  <GameOverlay center-content show-logo data-test="new">
    <template #title> Are you sure?</template>
    <template #msg>
      <p>This game will still count towards your statistics</p>
    </template>
    <template #buttons>
      <StartNewGameButton />
      <ContinueGameButton />
    </template>
  </GameOverlay>
</template>

<script>
import GameOverlay from '@/components/GameOverlay.vue';
import StartNewGameButton from '@/components/StartNewGameButton.vue';
import ContinueGameButton from '@/components/ContinueGameButton.vue';

export default {
  name: 'NewPage',
  components: {
    GameOverlay,
    StartNewGameButton,
    ContinueGameButton,
  },
};
</script>
