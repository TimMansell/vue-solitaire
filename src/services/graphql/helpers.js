export const formatResponse = (response) => ({
  error: false,
  response,
});

export const formatError = () => ({
  error: true,
});
