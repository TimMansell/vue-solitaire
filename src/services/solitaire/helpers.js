import shuffle from 'lodash.shuffle';

const shuffleCards = ({ values, suits }) => {
  const deck = values.flatMap((value, index) =>
    suits.map((suit) => {
      const card = {
        id: `${index}${suit}`,
        value,
        order: index + 1,
        suit,
        visible: false,
      };

      return card;
    })
  );

  const shuffledDeck = shuffle(deck);

  return shuffledDeck;
};

const showHideCards = (cards, offset = 0) =>
  cards.map((card, index) => {
    if ((index + offset) % 2 === 0) {
      return {
        ...card,
        visible: true,
      };
    }

    return card;
  });

const getSelectedCard = (cards, selectedCardId) => {
  const [selectedCard] = cards.flat().filter((card) => card.id === selectedCardId);

  return selectedCard;
};

const getSelectedCardPosition = (boardCards, selectedCardId) => {
  const columnNo = boardCards.findIndex((cards) =>
    cards.find((card) => card.id === selectedCardId)
  );

  const cardPosition = boardCards[columnNo].findIndex((card) => card.id === selectedCardId);

  return {
    columnNo,
    cardPosition,
  };
};

const getVisibleCards = (cards) => {
  const visibleCards = cards.flat().filter((card) => card.visible);

  return visibleCards;
};

const getLastCard = (board, selectedColumn) => {
  const [selectedCard] = board[selectedColumn].slice(-1);

  return selectedCard;
};

const getLastCards = (cards) => {
  const lastCards = cards.map((card) => card.slice(-1)).flat();

  return lastCards;
};

const moveCardsFrom = (selectedCardId, cards) => {
  const { columnNo, cardPosition } = getSelectedCardPosition(cards, selectedCardId);

  const columnCards = cards[columnNo].slice(0, cardPosition);

  const remainingCards = columnCards.map((card, index) => {
    if (index === columnCards.length - 1 && !card.visible) {
      const newValues = {
        ...card,
        visible: true,
      };
      return newValues;
    }

    return card;
  });

  return {
    column: columnNo,
    cards: remainingCards,
  };
};

const moveCardsTo = (selectedCardId, selectedColumn, cardsFrom, cardsTo) => {
  const { columnNo, cardPosition } = getSelectedCardPosition(cardsFrom, selectedCardId);

  const columnCards = cardsTo[selectedColumn];
  const moveCards = cardsFrom[columnNo].slice(cardPosition);

  const newColumn = [...columnCards, ...moveCards];

  return {
    column: selectedColumn,
    cards: newColumn,
  };
};

export {
  shuffleCards,
  showHideCards,
  getSelectedCard,
  getSelectedCardPosition,
  getLastCard,
  getLastCards,
  getVisibleCards,
  moveCardsFrom,
  moveCardsTo,
};
