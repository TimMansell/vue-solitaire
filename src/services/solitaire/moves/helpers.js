// eslint-disable-next-line import/prefer-default-export
export const displayMoves = (moves) => {
  if (moves.length) {
    console.log('---------------');
    moves.forEach(({ value, suit }) => {
      console.log('hasMove', `${value}${suit}`);
    });
  }
};
