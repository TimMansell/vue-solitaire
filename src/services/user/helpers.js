import { v4 as uuidv4 } from 'uuid';

export const getLocalUser = () => localStorage.getItem('luid');

export const createLocalUser = () => {
  const luid = uuidv4();

  localStorage.setItem('luid', luid);

  return luid;
};
