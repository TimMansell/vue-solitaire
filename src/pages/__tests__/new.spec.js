import { describe, it, expect } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import New from '@/pages/New.vue';
import { setupStore } from '@@/tests/helpers';

describe('New.vue', () => {
  it('renders the component without crashing', () => {
    const wrapper = shallowMount(New, {
      global: {
        mocks: {
          $store: setupStore({
            isOnline: true,
          }),
        },
      },
    });

    expect(wrapper.isVisible()).toBe(true);
  });
});
