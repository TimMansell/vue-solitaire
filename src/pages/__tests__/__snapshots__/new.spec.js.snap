// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New.vue matches snapshot 1`] = `
<gameoverlay-stub centercontent="true" showlogo="true" data-test="new"> Are you sure?<p>This game will still count towards your statistics</p>
  <startnewgamebutton-stub></startnewgamebutton-stub>
  <continuegamebutton-stub></continuegamebutton-stub>
</gameoverlay-stub>
`;
